WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Package Selection/Package Creation/ShippingLabelCustomPackageFormViewModel.swift

```diff
@@ -35,25 +37,114 @@ final class ShippingLabelCustomPackageFormViewModel: ObservableObject {
+
+    @Published private(set) var isNameValidated = true
+    @Published private(set) var isLengthValidated = true
+    @Published private(set) var isWidthValidated = true
+    @Published private(set) var isHeightValidated = true
+    @Published private(set) var isWeightValidated = true
+    private var isPackageValidated: Bool {
+        isNameValidated && isLengthValidated && isWidthValidated && isHeightValidated && isWeightValidated
+    }
+
+    // MARK: Initialization
+
+    init(package: ShippingLabelCustomPackage? = nil,
+         lengthUnit: String? = ServiceLocator.shippingSettingsService.dimensionUnit,
+         weightUnit: String? = ServiceLocator.shippingSettingsService.weightUnit) {
         self.lengthUnit = lengthUnit ?? ""
         self.weightUnit = weightUnit ?? ""
-        self.packageName = packageName
-        self.packageType = packageType
-        self.packageLength = packageLength
-        self.packageWidth = packageWidth
-        self.packageHeight = packageHeight
-        self.emptyPackageWeight = emptyPackageWeight
+        self.packageName = package?.title ?? ""
+        self.packageType = (package?.isLetter ?? false) ? .letter : .box
+        self.packageLength = package?.getLength().description ?? ""
+        self.packageWidth = package?.getWidth().description ?? ""
+        self.packageHeight = package?.getHeight().description ?? ""
+        self.emptyPackageWeight = package?.boxWeight.description ?? ""
+
+        configureFormValidation()
+    }
+}
+
+// MARK: - Validation
+extension ShippingLabelCustomPackageFormViewModel {
+
+    /// Validate each field on demand.
+    /// This ensures the package can be validated accurately even if one of the fields hasn't changed from its initial value.
+    ///
+    func validatePackage() {
+        isNameValidated = validatePackageName(packageName)
+        isLengthValidated = validatePackageDimension(packageLength)
+        isWidthValidated = validatePackageDimension(packageWidth)
+        isHeightValidated = validatePackageDimension(packageHeight)
+        isWeightValidated = validatePackageWeight(emptyPackageWeight)
+    }
+
+    /// Configure form validation, ignoring the initial value of each form field
+    ///
+    private func configureFormValidation() {
+        $packageName
+            .dropFirst()
+            .map { self.validatePackageName($0) }
+            .assign(to: &$isNameValidated)
+
+        $packageLength
+            .dropFirst()
+            .map { self.validatePackageDimension($0) }
+            .assign(to: &$isLengthValidated)
+
+        $packageWidth
+            .dropFirst()
+            .map { self.validatePackageDimension($0) }
+            .assign(to: &$isWidthValidated)
+
+        $packageHeight
+            .dropFirst()
+            .map { self.validatePackageDimension($0) }
+            .assign(to: &$isHeightValidated)
+
+        $emptyPackageWeight
+            .dropFirst()
+            .map { self.validatePackageWeight($0) }
```

Comment:
Since we mostly work with asynchronous code in Combine, I'd recommend not capturing `self` strongly to avoid retain cycles. You can read more about that [here](https://trycombine.com/posts/self-weak-unowned/).